test/Microsoft.VisualStudio.LanguageServices.Razor.Test/Editor/RazorDirectiveCompletionProviderTest.cs

```diff
@@ -0,0 +1,345 @@
+            await completionProvider.ProvideCompletionsAsync(context);
+
+            // Assert
+            var completionItems = GetCompletionItems(context);
+            var customDirectiveCompletion = Assert.Single(completionItems, item => item.DisplayText == customDirective.Directive);
+            AssertRazorCompletionItemDefaults(customDirectiveCompletion);
+            Assert.DoesNotContain(customDirectiveCompletion.Properties, property => property.Key == RazorDirectiveCompletionProvider.DescriptionKey);
+        }
+
+        private static void AssertRazorCompletionItem(string expectedDescription, CompletionItem item)
+        {
+            Assert.True(item.Properties.TryGetValue(RazorDirectiveCompletionProvider.DescriptionKey, out var actualDescription));
+            Assert.Equal(expectedDescription, actualDescription);
+
+            AssertRazorCompletionItemDefaults(item);
+        }
+
+        private static void AssertRazorCompletionItemDefaults(CompletionItem item)
+        {
+            Assert.Equal("_RazorDirective_", item.SortText);
+            Assert.False(item.Rules.FormatOnCommit);
+            var tag = Assert.Single(item.Tags);
+            Assert.Equal(CompletionTags.Intrinsic, tag);
+        }
+
+        private static RazorCodeDocumentProvider CreateCodeDocumentProvider(string text, IEnumerable<DirectiveDescriptor> directives)
+        {
+            var codeDocumentProvider = new Mock<RazorCodeDocumentProvider>();
+            var codeDocument = TestRazorCodeDocument.CreateEmpty();
+            var sourceDocument = TestRazorSourceDocument.Create(text);
+            var options = RazorParserOptions.Create(builder =>
+            {
+                foreach (var directive in directives)
+                {
+                    builder.Directives.Add(directive);
+                }
+            });
+            var syntaxTree = RazorSyntaxTree.Parse(sourceDocument, options);
+            codeDocument.SetSyntaxTree(syntaxTree);
+            codeDocumentProvider.Setup(provider => provider.TryGetFromDocument(It.IsAny<TextDocument>(), out codeDocument))
+                .Returns(true);
+
+            return codeDocumentProvider.Object;
+        }
+
+        private static CompletionContext CreateContext(int position, RazorDirectiveCompletionProvider completionProvider, Document document)
+        {
+            var context = new CompletionContext(
+                completionProvider,
+                document,
+                position,
+                TextSpan.FromBounds(position, position),
+                CompletionTrigger.Invoke,
+                new Mock<OptionSet>().Object,
+                CancellationToken.None);
+
+            return context;
+        }
+
+        private IReadOnlyList<CompletionItem> GetCompletionItems(CompletionContext context)
+        {
+            var completionItems = (IReadOnlyList<CompletionItem>)_completionContextItemsField.GetValue(context);
+
+            return completionItems;
```

Comment:
What if you just refactored the code to be more testable rather than this?

Instead of calling `ProvideCompletionsAsync` in tests, make an internal `GetCompletionsAsync` and call that?

Reply:
Sure, will have to be virtual in order to allow for detection of "attempted to get completions"